### PR TITLE
Implement formal data structure for memory type

### PR DESCRIPTION
**Pull request by pipermerriam**
Builds from #22 

## What was wrong?

The spec defines *memory type* as formal data structure.

https://webassembly.github.io/spec/core/bikeshed/index.html#memory-types%E2%91%A0

This type is identical to the *limits* type from #21

## How was it fixed?

Implemented an equivalent data type to the limits one.  Intentionally did not subclass limits to ensure that any kind of instance checks between the different types would beep them separate.

Also resolved a `TODO` for how the validation of a limits object should happen.

#### Cute Animal Picture

surprised-black-spider-monkey-funny-animal-picture

### CODE REVIEW

File: wasm/main.py [hunk 1]
```diff
@@ -196,27 +201,17 @@ def spec_validate_functype(ft):
 
 
 def spec_validate_tabletype(table_type: TableType) -> TableType:
-    # TODO: use of UINT32_CEIL may be incorrect here as `validate_limit` checks
-    # against the value using `>`.  Check looks like it **should** be using
-    # `>=` to ensure that the limit bounds fit withint a UINT32 but updating
-    # this to use `UINT32_MAX` is currently causing spec test failures.
-    # TODO: The type casting to `UInt32` is technically incorrect here since
-    # `UINT32_CEIL` is outside the valid range for u32, but we first need to
-    # resolve the above TODO.
```
Comment (pipermerriam) [inline wasm/main.py]: cc @carver I resolved this.

The `k` variable, now referred to as `upper_bound` is intended to be an inclusive upper bound as per the spec, but **also** there was some potentially missing validation on the general constraints of the `limits.max` and `limits.min` that I added to `spec_validate_limits` but it's also worth noting that this validation is duplicated in [`spec_binary_uN`](https://github.com/ethereum/py-wasm/blob/2fa5a06d59d456053002cfc79bf7bb71f93c8245/wasm/main.py#L3318-L3325) which is used to extract the values that become `limits.min` and `limits.max` so at some point this should be consolidated.

File: wasm/main.py [hunk 2]
```diff
@@ -196,27 +201,17 @@ def spec_validate_functype(ft):
...
 
 
-def spec_validate_memtype(limits: Limits) -> Limits:
-    # TODO: use of UINT32_CEIL may be incorrect here as `validate_limit` checks
-    # against the value using `>`.  Check looks like it **should** be using
-    # `>=` to ensure that the limit bounds fit withint a UINT32 but updating
-    # this to use `UINT32_MAX` is currently causing spec test failures.
+def spec_validate_memtype(memory_type: MemoryType) -> MemoryType:
+    limits = Limits(memory_type.min, memory_type.max)
     spec_validate_limit(limits, constants.UINT16_CEIL)
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply wasm/main.py]: No, the validation is correctly using the `CEIL` version.  When validating limits, if the limit is supposed to be validated against an upper bound of `2**16` then it is correct to pass in `2**16` and it is correct for this validation function to do the check as `assert m <= upper_bound`, aka `assert m <= 2**16`.

Formal definition here: https://webassembly.github.io/spec/core/bikeshed/index.html#limits%E2%91%A2

[the remark this answers is not in the document]
Comment (pipermerriam) [reply wasm/main.py]: You are correct that `UINT16_CEIL` doesn't fit inside of an unsigned 16-bit integer, however, the spec for how validation is supposed to occur isn't defined in terms of integer bit sizes, only a scalar value with no bound, and the validation definition is inclusive of this upper bound.

[the remark this answers is not in the document]
Comment (carver) [reply wasm/main.py]: Huh, very odd choice on the part of the spec. But I see the ceiling now after digging into the memtype. :+1:

**Pull request merged**